### PR TITLE
KAFKA-17018: update MetadataVersion for the Kafka release 3.9

### PR DESCRIPTION
**Pull request by cmccabe**
- Mark 3.9-IV0 as stable. Metadata version 3.9-IV0 should return Fetch version 17.

- Move ELR to 4.0-IV0. Remove 3.9-IV1 since it's no longer needed.

- Create a new 4.0-IV1 MV for KIP-848.

### CODE REVIEW

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java [hunk 1]
```diff
@@ -244,7 +244,7 @@ public enum MetadataVersion {
      * <strong>Think carefully before you update this value. ONCE A METADATA VERSION IS PRODUCTION,
      * IT CANNOT BE CHANGED.</strong>
      */
-    public static final MetadataVersion LATEST_PRODUCTION = IBP_3_8_IV0;
+    public static final MetadataVersion LATEST_PRODUCTION = IBP_3_9_IV0;
```
Comment (junrao) [inline server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: Hmm, ListOffsetRequest has latestVersionUnstable set to true. So, it's kind of weird to have a production MV depending on a unstable PRC.

@clolov @showuon @chia7712 : Is KIP-1005 done? Could we set latestVersionUnstable to false for ListOffsetRequest in 3.9?

File: server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java [hunk 2]
```diff
@@ -23,8 +23,8 @@ public enum TestFeatureVersion implements FeatureVersion {
     TEST_0(0, MetadataVersion.MINIMUM_KRAFT_VERSION, Collections.emptyMap()),
     // TEST_1 released right before MV 3.7-IVO was released, and it has no dependencies
     TEST_1(1, MetadataVersion.IBP_3_7_IV0, Collections.emptyMap()),
-    // TEST_2 released right before MV 3.9-IVO was released, and it depends on this metadata version
-    TEST_2(2, MetadataVersion.IBP_3_9_IV0, Collections.singletonMap(MetadataVersion.FEATURE_NAME, MetadataVersion.IBP_3_9_IV0.featureLevel()));
+    // TEST_2 released right before MV 4.0-IVO was released, and it depends on this metadata version
+    TEST_2(2, MetadataVersion.IBP_4_0_IV0, Collections.singletonMap(MetadataVersion.FEATURE_NAME, MetadataVersion.IBP_4_0_IV0.featureLevel()));
```
Comment (junrao) [inline server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: @jolshan : What MV should TEST_2 depend on? Do we need to keep changing it with new MVs? It would be useful to add a comment to document this.

File: metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java [hunk 3]
```diff
@@ -389,7 +389,7 @@ public void testNoLeaderEpochBumpOnIsrExpansionDuringMigration(String metadataVe
      * always results in a leader epoch increase.
      */
     @ParameterizedTest
-    @ValueSource(strings = {"3.4-IV0", "3.5-IV2", "3.6-IV0", "3.7-IV2", "3.9-IV1"})
+    @ValueSource(strings = {"3.4-IV0", "3.5-IV2", "3.6-IV0", "3.7-IV2", "4.0-IV0"})
```
Comment (junrao) [inline metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java:392]: I had a followup comment in https://github.com/apache/kafka/pull/16347#discussion_r1661558720. Do we need to add 4.0-IV0 for `testNoLeaderEpochBumpOnEmptyTargetIsr`?

**Review by junrao — COMMENTED** [review]
@cmccabe : Thanks for the PR. Left a few comments.

File: server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java (hunk 2 above)
Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: I'm sorry I haven't gotten a chance to update this. I was hoping to have one of the test versions be the latest unstable version, but I believe I updated the tests for them to work either way now. Let me double check. 

If so we can just keep at 3.9.

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: Ok. This is the test that will fail if we don't change the version to 4.0. https://github.com/apache/kafka/blame/98cdf9717049e87ba34bb5161276577fcb8bd1c4/server-common/src/test/java/org/apache/kafka/server/common/FeaturesTest.java#L137

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: We have a similar test here for real features, so if we think this is sufficient, we can remove the test I linked above. https://github.com/apache/kafka/blame/98cdf9717049e87ba34bb5161276577fcb8bd1c4/server-common/src/test/java/org/apache/kafka/server/common/FeaturesTest.java#L51

Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: Thanks, Justine. Could we just use `MetadataVersion.latestTesting()` to avoid having to keep updating this with new MVs?

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: LatestTesting doesn't quite work since it is not guaranteed that latestTesting is unstable right?

Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: Well, the only case when LatestTesting is stable is that there is no unstable MV, right?

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: Right -- in that case this test will fail. https://github.com/apache/kafka/blame/98cdf9717049e87ba34bb5161276577fcb8bd1c4/server-common/src/test/java/org/apache/kafka/server/common/FeaturesTest.java#L137

My understanding is that we don't create a new MV when we mark one as stable, so for example, if all the MVs were marked stable, the test would fail.

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: Alternatively, I could write the test to check if the latest MV is stable, and if so, skip the check.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java (hunk 1 above)
Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > Is KIP-1005 done? Could we set latestVersionUnstable to false for ListOffsetRequest in 3.9?

KIP-1005 is completed and backport to 3.9, so yes latestVersionUnstable should be false now.

Comment (cmccabe) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: @junrao : yes, KIP-1005 will ship in AK 3.9.

File: metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java (hunk 3 above)
Comment (cmccabe) [reply metadata/src/test/java/org/apache/kafka/controller/PartitionChangeBuilderTest.java:392]: It would not make sense to add 4.0-IV0 to `testNoLeaderEpochBumpOnEmptyTargetIsr`, since in 4.0 an empty target ISR actually WILL trigger a leader epoch bump.

The reason is because once KIP-966 is implemented, empty ISRs will actually be possible, whereas they are not prior to KIP-966

File: server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java (hunk 2 above)
Comment (cmccabe) [reply server-common/src/main/java/org/apache/kafka/server/common/TestFeatureVersion.java:27]: We should probably just document that the last MV should always be unstable. I don't think it's ever been stable since "unstable" was created (we don't seem to stop developing features...)

Should we do this in a follow-on so that we can unblock 3.9?

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java (hunk 1 above)
Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: @chia7712 do you plan to update https://github.com/apache/kafka/blob/df04887ba597941674a28d54bd35202bbd21631e/clients/src/main/resources/common/message/ListOffsetsRequest.json#L42

Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > do you plan to update 

I assumed @cmccabe will update it in this PR :smile: 

Will file a jira and PR for trunk and 3.9 later

**Review by junrao — APPROVED** [review]
@cmccabe : Thanks for the explanation. LGTM assuming we will have a followup PR to address https://github.com/apache/kafka/pull/16841#discussion_r1711938781.

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java (hunk 1 above)
Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: Thanks, @chia7712. It's a bit surprising that no test failed with a production MV depending on an unstable ListOffset RPC. Do you know why?

**Comment by chia7712** [comment]
> Thanks for the explanation. LGTM assuming we will have a followup PR to address https://github.com/apache/kafka/pull/16841#discussion_r1711938781.

will file PR later (https://issues.apache.org/jira/browse/KAFKA-17319)

File: server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java (hunk 1 above)
Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > It's a bit surprising that no test failed with a production MV depending on an unstable ListOffset RPC. Do you know why?

I don't think there is the "connection" between MV and unstable ListOffset RPC. The only condition used to enable unstable RPC is to set `unstable.api.versions.enable=true` regardless of MV.

For another, the version supports of ListOffset is a bit weird. We don't add the version check on both client/server side for the "specific" timestamp. 

For example: version 8 (KIP-405) added `EARLIEST_LOCAL_TIMESTAMP=-4` flag. the question are: 

1. Should server return `UNSUPPORTED_VERSION` if server receive a request with `version<8` and `timestamp=-4`?
2. Should admin check the timestamp and throw unsupported version if the server is unsupported to that timestamp flag?

It seems to me the version check is necessary to make 1) newer server can return suitable error to older client and 2) newer client can generate suitable error itself rather than receive undefined result from older server. 

@junrao WDYT?

Comment (showuon) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: I agree we should add tests for it.

Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: @chia7712 : Thanks for the reply. Previously, Christo's PR seems to have caused some tests to fail because of this (https://github.com/apache/kafka/pull/15673#issuecomment-2104306940). So, I was wondering why the failure didn't show up here.

Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > Christo's PR seems to have caused some tests to fail because of this (https://github.com/apache/kafka/pull/15673#issuecomment-2104306940). So, I was wondering why the failure didn't show up here.

the CI is gone so I can't see "which test" you discussed. According to https://github.com/apache/kafka/commit/53276b12c57f325fc62e0133eae1eae023c3b74b,  I guess the test case is `testApiVersionsRequestIncludesUnreleasedApis`. It is used to verify the api key and version range. Everything is good but #15761 introduced a "wrong" config to it. However, the test was not failed since we did not have "unstable" version yet.

That is why #15673 changed the `unstable.api.versions.enable` from `false` to `true`, because it added a unstable "version=0" to LIST_OFFSET.

In short, I feel the test case `testApiVersionsRequestIncludesUnreleasedApis` is NOT related to MV. If we want to build the relationship between "unstable version" and "MV". Maybe we can add new field to `ApiKeys` to declare the MV for the RPC having "unstable version". And then we can add tests for following checks

1. the RPC having unstable version must declare a MV in `ApiKeys`. The MV is equal to the release which can change the version from unstable to "stable"
2. the `MetadataVersion#LATEST_PRODUCTION` must be smaller than MV in `ApiKeys`. That ensures `LATEST_PRODUCTION` will NOT depend on unstable versions.

This idea does NOT impact the production code, but it increase the complexity of updating MV/version.

@junrao WDYT?

Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > For example: version 8 (KIP-405) added EARLIEST_LOCAL_TIMESTAMP=-4 flag. the question are:
    Should server return UNSUPPORTED_VERSION if server receive a request with version<8 and timestamp=-4?
    Should admin check the timestamp and throw unsupported version if the server is unsupported to that timestamp flag?
It seems to me the version check is necessary to make 1) newer server can return suitable error to older client and 2) newer client can generate suitable error itself rather than receive undefined result from older server.

I open https://issues.apache.org/jira/browse/KAFKA-17331 for it. @cmccabe @junrao Sorry that I feel this is a part of KIP-1005. we will complete it ASAP

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: I don't think MV needs to tie to unstable versions. We used to not have unstable metadata versions in a release, but now we allow it if the MV is for the next release and we have the mechanisms to do this. 

I think generally, the unstable api version in a release is OK, BUT if we say the KIP is completed in the release, the api should be stable (unless we doing using early access etc). I don't think we need to tie api versions to MV at this time.

Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: >  I don't think we need to tie api versions to MV at this time.

@jolshan : Just to be clear. I think the issue is that some inter broker requests (e.g. replication) depend on MV to set the request version, not based on supported versions returned in ApiVersion. If we have a stable MV depending on an unstable version of an inter-broker RPC, in the default setting (with stable MV and requests), the broker will receive a newer version of the request that it doesn't support.

@chia7712 : For this particular case, `ListOffset` is an inter broker RPC. In the default setting, the replication code will use v9 of `ListOffset`. However, by default, v9 of `ListOffset` is not supported with latestVersionUnstable set to true. So, I still don't understand why some existing tests don't fail. We probably need to understand the testing gap. If we have sufficient test coverage, we probably don't need additional protocols.

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: Yes -- sorry I was imprecise. If the api version is advertised by a MV that is stable it sould also be stable.

I didn't realize ListOffset was being used in the replication path. I guess this is just for tiered storage.

Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > However, by default, v9 of ListOffset is not supported with latestVersionUnstable set to true. So, I still don't understand why some existing tests don't fail. We probably need to understand the testing gap. If we have sufficient test coverage, we probably don't need additional protocols.

yes, you are right. we should have a test for replica path. I just write a IT which can produce following error when LIST_OFFSET has unstable version:
```
[2024-08-14 03:27:13,119] ERROR Closing socket for 127.0.0.1:36885-127.0.0.1:48854-3 because of error (kafka.network.Processor:76)
org.apache.kafka.common.errors.InvalidRequestException: Received request api key LIST_OFFSETS with version 9 which is not enabled
```
Also, it seems we create IT with unstable APIs always [0] and I guess that is why we don't see failed tests.

**Summary**
1. broker-to-broker: the LIST_OFFSET version is based on MV [1], and we use IT to protect us from updating MV with an unstable version (https://issues.apache.org/jira/browse/KAFKA-17336)
2. new-client-to-old-broker: the builder of LIST_OFFSET will check the timestamp and its matched version (https://issues.apache.org/jira/browse/KAFKA-17331)
3. old-client-to-new-broker: the broker will check the timestamp and its matched version (https://issues.apache.org/jira/browse/KAFKA-17331)


> I didn't realize ListOffset was being used in the replication path. I guess this is just for tiered storage.

@jolshan not really, the request is used to truncate in the replication path. I have attached a sample code in https://issues.apache.org/jira/browse/KAFKA-17336. we will complete it tomorrow.

[0] https://github.com/apache/kafka/blob/8d29bc1fa8ce75b6807b957ea35ad39e8dd9e04a/core/src/test/scala/unit/kafka/utils/TestUtils.scala#L307
[1] https://github.com/apache/kafka/blob/8d29bc1fa8ce75b6807b957ea35ad39e8dd9e04a/server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java#L474

Comment (jolshan) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: Thanks. https://issues.apache.org/jira/browse/KAFKA-17336 should be good to not release with unstable versions that break api usage.

Comment (junrao) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > Also, it seems we create IT with unstable APIs always [0] and I guess that is why we don't see failed tests.

Thanks for the investigation, @chia7712. Maybe we just need to disable unstable API/MV in a few existing integration tests.

Comment (chia7712) [reply server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java:247]: > Maybe we just need to disable unstable API/MV in a few existing integration tests.

yep, we are looking for the similar tests
